northstar/src/runtime/cgroups.rs

```diff
@@ -63,104 +134,144 @@ impl CGroups {
+    let mut event_control = fs::OpenOptions::new()
+        .write(true)
+        .open(&event_control)
+        .await
+        .expect("Failed to open event_control");
+    event_control
+        .write_all(format!("{} {}", event_fd.as_raw_fd(), oom_control.as_raw_fd()).as_bytes())
+        .await
+        .expect("Failed to setup event_control");
+    event_control
+        .flush()
+        .await
+        .expect("Failed to setup oom event fd");
 
     // This task stops when the main loop receiver closes
-    task::spawn(async move {
-        let mut interval = time::interval(time::Duration::from_millis(500));
-        loop {
+    let task = {
+        let stop = stop.clone();
+        task::spawn(async move {
+            debug!("Listening for oom events of {}", container);
+            let mut buffer = [0u8; 1];
+
             select! {
-                _ = stop.cancelled() => break,
-                _ = tx.closed() => break,
-                _ = interval.tick() => {
-                    match fs::read_to_string(&oom_control).await {
-                        Ok(s) => {
-                            if s.lines().any(|l| l == UNDER_OOM) {
-                                warn!("Container {} is under OOM!", container);
-                                tx.send(Event::Oom(container)).await.ok();
-                                break;
-                            }
-                        }
-                        Err(_) => {
-                            debug!("Stopping oom monitor of {}", container);
-                            break;
+                _ = stop.cancelled() => (),
+                _ = tx.closed() => (),
+                _ = event_fd.read(&mut buffer) => {
+                    loop {
+                        match tx.try_send(Event::Oom(container.clone())) {
+                            Ok(_) => break,
+                            Err(TrySendError::Closed(_)) => break,
+                            Err(TrySendError::Full(_)) => time::sleep(time::Duration::from_millis(1)).await,
                         }
                     }
                 }
             }
-        }
-    });
-
-    Ok(())
-}
-
-async fn write(path: &Path, value: &str) -> Result<(), Error> {
-    fs::write(path, value)
-        .await
-        .map_err(|e| Error::Io(format!("Failed to write to {}", path.display()), e))
+            drop(event_fd);
+            drop(oom_control);
+            drop(event_control);
+            debug!("Stopped oom monitor of {}", container);
+        })
+    };
+    (stop, task)
 }
 
 /// Get the cgroup v1 controller hierarchy mount point
 fn mount_point(controller: &str) -> Result<PathBuf, Error> {
-    let controller = controller.to_owned();
-    MountIter::new()
-        .map_err(Error::MountInfo)?
-        .filter_map(Result::ok)
-        .find(|m| m.fstype == "cgroup" && m.options.contains(&controller))
-        .map(|m| m.dest)
-        .ok_or(Error::UnknownController(controller))
+    task::block_in_place(|| {
```

Comment:
is this needed even if the function is not `async`?

Reply:
In this case yes because the fn is called from an async context (as anything inside northstar). I'd wish there would be some kind of annotation of in best case automatic detection of blocking things. Think the async-std guys tried something like that.